primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -290,6 +292,89 @@ impl<E: Pairing> PolynomialCommitmentScheme for UnivariateKzgPCS<E> {
+    pub fn batch_open_fk23(
+        prover_param: impl Borrow<UnivariateProverParam<E>>,
+        polynomial: &<Self as PolynomialCommitmentScheme>::Polynomial,
+        points: BatchEvalPoints<E::ScalarField>,
+    ) -> Result<
+        (
+            Vec<<Self as PolynomialCommitmentScheme>::Proof>,
+            Vec<<Self as PolynomialCommitmentScheme>::Evaluation>,
+        ),
+        PCSError,
+    > {
+        let d = polynomial.degree();
+        if !d.is_power_of_two() {
+            return Err(PCSError::InvalidParameters(
+                "FK23 only supports polynomial with power-of-two degree".to_string(),
+            ));
+        }
+
+        // Step 1. compute \vec{h} using fast Toeplitz matrix multiplication
+        // 1.1 Toeplitz matrix A (named `poly_coeff_matrix` here)
+        let mut toep_col = vec![*polynomial
+            .coeffs()
+            .last()
+            .ok_or_else(|| PCSError::InvalidParameters("poly degree should >= 1".to_string()))?];
+        toep_col.resize(d, <<E as Pairing>::ScalarField as Field>::ZERO);
+        let toep_row = polynomial.coeffs().iter().skip(1).rev().cloned().collect();
+        let poly_coeff_matrix = ToeplitzMatrix::new(toep_col, toep_row)?;
+
+        // 1.2 vector s (named `srs_vec` here)
+        let srs_vec: Vec<E::G1> = prover_param
+            .borrow()
+            .powers_of_g
+            .iter()
+            .take(d)
+            .rev()
+            .cloned()
+            .map(|g| g.into_group())
+            .collect();
+
+        // 1.3 compute \vec{h}
+        let h_vec = poly_coeff_matrix.fast_vec_mul(&srs_vec)?;
+
+        // Step 2. evaluate h(X) with coeffs \vec{h} at `points` using FFT
+        let mut h_poly = GeneralDensePolynomial::from_coeff_vec(h_vec);
+        // NOTE!! this is the trickist subtle line in this function !!
+        // `polynomial` has a degree `d` which is guaranteed to be power-of-two to use
+        // Toeplitz fast mul algorithm. h(X) has a degree `d-1` which cause a plain
+        // domain instantiation on `h_poly` half the size of that for `polynomial`,
+        // which leads to different vector of roots of unity used, ultimately leading to
+        // wrong proofs. Thus, to make sure during FFT, the same vector of roots of
+        // unity are used, we need to manually resize the coefficients so that
+        // `GeneralDensePolynomial::batch_evaluate` are informed to use the domain of
+        // size (d+1).next_power_of_two().
+        h_poly.coeffs.resize(polynomial.coeffs.len(), E::G1::zero());
+
+        let proofs: Vec<_> = h_poly
+            .batch_evaluate(points)
+            .into_iter()
+            .map(|g| UnivariateKzgProof {
+                proof: g.into_affine(),
+            })
+            .collect();
+
+        // Evaluate at all points
+        ark_std::println!("!! during eval:");
+        let evals =
+            GeneralDensePolynomial::from_coeff_slice(&polynomial.coeffs).batch_evaluate(points);
```

Comment:
It's not true that we only support roots-of-unity. (see the tests, I have both general points and roots)

When we are using general points, teh evaluations are done using Horner methods on each individual points, no FFT used. The computation of proofs (`h(X)`) has no change.

Reply:
API question: is it better to have a single function take an enum argument that instructs whether to use given (arbitrary) points vs. roots of unity (like you have now), or should we instead split it into two calls:
```rust
batch_open(arbitrary_points_slice)
batch_open_roots_of_unity(log_num_points)
```

Reply:
lemme try to explain my hesitance:

"batching" could mean multi-/single-poly evaluated at multi-/single points (at least 4 combinations, I'm not even counting the "multi-eval-domain" version in HaloInfinite).
If we a single API that accepts multi-poly, and multi-points (which is what we currently defined in the `trait PCS::batch_open()`, while we can make the decision inside concrete instantiation what's the default techniques we use. (e.g. for single-poly multi-points, do we use KZG10's idea which renders shorter proof but slower prover, or do we use FK23 whose proof size is linear in number of points, but faster prover). 
Ultimately, we want the caller/user to express their preference of which batching they use in their specific use cases. 
This can be done in two ways. Either we accept an additional argument (an optional enum: `batch_open(..., flavor: Option<BatchingFlavor>)`; or we simply separate out each batching into its own functions, making it more explicit choice and don't over-abstract the common trait interfaces.
(^^ and I went for the latter)

coming back to the two APIs @ggutoski proposed. 
Firstly, FK23 can deal with both roots-of-unity and general points, so I'm not sure it's the best ergonomic to separate them out into two functions.

I guess the question is do we want *one function per batching technique, or one function per batching pre-condition (only roots-of-unity, or only single-poly-multi-point)*?

Reply:
will continue in https://github.com/EspressoSystems/jellyfish/pull/231#discussion_r1173958303.